certbot/tests/error_handler_test.py

```diff
@@ -26,34 +46,51 @@ def test_context_manager(self):
             with self.handler:
                 raise ValueError
         except ValueError:
-            pass
+            exception_raised = True
+
+        self.assertTrue(exception_raised)
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
 
-    @mock.patch('certbot.error_handler.os')
-    @mock.patch('certbot.error_handler.signal')
-    def test_signal_handler(self, mock_signal, mock_os):
-        # pylint: disable=protected-access
-        mock_signal.getsignal.return_value = signal.SIG_DFL
-        self.handler.set_signal_handlers()
-        signal_handler = self.handler._signal_handler
-        for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal_handler)
+    def test_context_manager_with_signal(self):
+        with signal_receiver(self.signals) as signals_received:
+            with self.handler:
+                should_be_42 = 42
+                send_signal(self.signals[0])
+                should_be_42 *= 10
 
-        signum = self.signals[0]
-        signal_handler(signum, None)
+        # check exectuion stoped when the signal was sent
+        self.assertEqual(42, should_be_42)
+        # assert signals were caught
+        self.assertEqual([self.signals[0]], signals_received)
+        # assert the error handling function was just called once
         self.init_func.assert_called_once_with(*self.init_args,
                                                **self.init_kwargs)
-        mock_os.kill.assert_called_once_with(mock_os.getpid(), signum)
-
-        self.handler.reset_signal_handlers()
         for signum in self.signals:
-            mock_signal.signal.assert_any_call(signum, signal.SIG_DFL)
+            sig = signal.getsignal(signum)
+            self.assertTrue((sig == signal.SIG_DFL) or (sig == signal.SIG_IGN))
```

Comment:
That sounds like a good solution.

I did this little hack because `getsignal(SIGXFSZ)` returns `SIG_IGN`. But before `getsignal` was mocked to return `SIG_DFL`.